### PR TITLE
[Ruby] do-end Blocks (with implicit &Proc and yield handling) #3928

### PR DESCRIPTION
**Pull request by badly-drawn-wizards**
- Adds yield RubyNode
- Handles yield in RubyNodeCreator
- Generates a ProcParameter with a unique identifier if a yield is used in a method with no block parameter
- Generates a call to the block parameter for yields

### CODE REVIEW

File: joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstForExpressionsCreator.scala [hunk 1]
```diff
@@ -365,6 +369,19 @@ trait AstForExpressionsCreator(implicit withSchemaValidation: ValidationMode) {
     astForSimpleCall(node.asSimpleCall)
   }
 
+  protected def astForYield(node: YieldExpr): Ast = {
+    scope.useProcParam match {
+      case Some(param) =>
+        astForMemberCall(
+          MemberCall(SimpleIdentifier()(node.span.spanStart(param)), ".", "call", node.arguments)(node.span)
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstForExpressionsCreator.scala:376]: Could we lower this differently? My concern is that this is closer to reflection in how this is executed, and not sure that our call graph could resolve this.

Could this instead be a `SimpleCall`?

File: joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstForFunctionsCreator.scala [hunk 2]
```diff
@@ -76,12 +79,18 @@ trait AstForFunctionsCreator(implicit withSchemaValidation: ValidationMode) { th
       astForMethodBody(node.body, optionalStatementList)
     }
 
+    val anonProcParam = scope.anonProcParam.map { param =>
+      val paramNode = ProcParameter(param)(node.span.spanStart(s"&$param"))
+      val index = parameterAsts.lastOption.flatMap(_.root).map { case m: NewMethodParameterIn => m.index }.getOrElse(0)
+      astForParameter(paramNode, index)
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstForFunctionsCreator.scala:85]: Should this not add one?
```suggestion
      astForParameter(paramNode, index + 1)
```

File: joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/datastructures/RubyScope.scala [hunk 3]
```diff
@@ -117,6 +118,42 @@ class RubyScope(summary: RubyProgramSummary, projectRoot: Option[String])
...
+    case scope @ ScopeElement(MethodScope(fullName, param, _), variables) =>
+      (ScopeElement(MethodScope(fullName, param, true), variables), param.fold(x => x, x => x))
+  }
+
+  def anonProcParam: Option[String] = stack.collectFirst { case ScopeElement(MethodScope(_, Left(param), true), _) =>
+    param
+  }
+
+  def setProcParam(param: String): Unit = updateSurrounding {
+    case scope @ ScopeElement(MethodScope(fullName, _, _), variables) =>
```
Comment (badly-drawn-wizards) [inline joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/datastructures/RubyScope.scala:153]: TODO remove unused `scope @`s

File: joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstForFunctionsCreator.scala [hunk 4]
```diff
@@ -268,8 +290,20 @@ trait AstForFunctionsCreator(implicit withSchemaValidation: ValidationMode) { th
 
         val stmtBlockAst = astForMethodBody(node.body, optionalStatementList)
 
+        val anonProcParam = scope.anonProcParam.map { param =>
+          val paramNode = ProcParameter(param)(node.span.spanStart(s"&$param"))
+          val index =
+            parameterAsts.lastOption.flatMap(_.root).map { case m: NewMethodParameterIn => m.index }.getOrElse(1)
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstForFunctionsCreator.scala:296]: This `getOrElse` is at `1`, while the above is at `0`, is this intentional? I would think, on the condition that you should append `1` to index at `astForParameter`, that the truth should be `0`

File: joern-cli/frontends/rubysrc2cpg/src/test/scala/io/joern/rubysrc2cpg/querying/ProcParameterAndYieldTests.scala [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+          forAll(cpgs)(_.call.code("yield").receiver.isCall.argument.code.l shouldBe List("<proc-param-0>", "call"))
+        }
+
+        "add a block argument" in {
+          forAll(cpgs)(_.method("foo").parameter.code("&.*").name.l shouldBe List("<proc-param-0>"))
+        }
+      }
+    }
+    "that don't have a yield nor a proc parameter" should {
+      val cpg1 = code("""| def foo() end
```
Comment (DavidBakerEffendi) [inline joern-cli/frontends/rubysrc2cpg/src/test/scala/io/joern/rubysrc2cpg/querying/ProcParameterAndYieldTests.scala:41]: These code blocks are single line, perhaps we could make it more concise with a normal string? e.g. `code("def foo() end")`

File: joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstForExpressionsCreator.scala (hunk 1 above)
Comment (DavidBakerEffendi) [reply joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstForExpressionsCreator.scala:376]: Also, concerning flows, the yield needs to map to `MethodReturn`, how is this done in `pysrc2cpg`/should we add a `return` here?

**Review by DavidBakerEffendi — CHANGES_REQUESTED** [review]

File: joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstForFunctionsCreator.scala (hunk 4 above)
Comment (badly-drawn-wizards) [reply joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstForFunctionsCreator.scala:296]: I should add that to the test.

Comment (badly-drawn-wizards) [reply joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstForFunctionsCreator.scala:296]: But the off by one error is not.

File: joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstForExpressionsCreator.scala (hunk 1 above)
Comment (badly-drawn-wizards) [reply joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstForExpressionsCreator.scala]:
```scala
      case node: ast.Yield          => unhandled(node)
      case node: ast.YieldFrom      => unhandled(node)
```
it isn't handled there, but they would need to be handled vastly differently anyway.

**Review by DavidBakerEffendi — APPROVED** [review]
Looks good, let's follow your suggestion (from our chat offline) to add some kind of custom conditional to handle the case where yields can return or generate a collection, etc.

**Pull request merged**